src/cpp/src/text_streamer.cpp

```diff
@@ -122,7 +132,73 @@ void TextStreamer::end() {
     return;
 }
 
-ov::genai::StreamerBase::~StreamerBase() = default;
+StreamerBase::~StreamerBase() = default;
+
+// Is used to hide internal states of TextParserStreamer
+class TextParserStreamer::TextParserStreamerImpl {
+public:
+
+std::vector<std::shared_ptr<IncrementalParser>> m_parsers;
+JsonContainer m_parsed_message;
+
+TextParserStreamerImpl(std::vector<std::shared_ptr<IncrementalParser>> parsers) : m_parsers{parsers} {}
+};
+
+TextParserStreamer::TextParserStreamer(const Tokenizer& tokenizer, std::vector<std::shared_ptr<IncrementalParser>> parsers) 
+    : TextStreamer(tokenizer, [this](std::string s) -> CallbackTypeVariant {
+                return this->write(s);
+    }), m_pimpl{std::make_unique<TextParserStreamerImpl>(parsers)} {}
+
+CallbackTypeVariant TextParserStreamer::write(std::string message) {
+    // When 'write' is called with string, it means new chunck of tokens is decoded into text
+
+    auto flushed_tokens = std::vector<int64_t>();
+    if (message.back() == '\n') {
+        // Flush all tokens // TODO: m_decoded_lengths[m_decoded_lengths.size() - 1] = -1;
+        flushed_tokens.assign(m_tokens_cache.begin(), m_tokens_cache.end());
+    } else if (m_decoded_lengths.size() >= delay_n_tokens) {
+        // prompt          = "I was waiting for the bus.\n"
+        // tokens          = [2,2,  3,      45, 67, 89,4,2]
+        // decoded_lengths = [1,5,  13,     17, 21, 25,26,27]
+        // let printed_len = 13 (after "I was waiting")
+        // then delta_text = "for the bus.\n"
+        // delta_tokens = [45, 67, 89,4,2]
+        // delta_tokens = m_tokens_cache[4..end]
+
+        // Find where the last printed tokens are located based on m_printed_len and print_until
+        auto print_until = m_decoded_lengths[m_decoded_lengths.size() - delay_n_tokens];
```

Comment:
Can we reuse `m_printed_len` instead of recalculating `print_untill` here?

We need to swap the callback call in `TextStreamer::write(token_id)`:
```cpp
    auto status = run_callback_if_needed(res.str());
    
    if (print_until > -1 && print_until > m_printed_len) {
        m_printed_len = print_until;
    }
    return status;
```